### ISSUE TITLE
Get away from generating C++ files in python

### ISSUE THREAD

**Issue by Jupeyy** [ddnet/ddnet#5648]
I think it's not really nice style, IDE's find the wrong include and that's confusing. We can do most stuff with constexpr or (c++20) consteval.

**Comment by Chairn** [comment]
Would be nice yup. The python stuff is kinda weird. New addition can simply be done by copy pasting existing net structures and associated function. Would also ease to add some function to the net structures (like copy operator to remove unsafe mem_copy).

**Comment by heinrich5991** [comment]
Needs some discussion.

The protocol generation has upsides, namely that one cannot accidentally write unsafe decoding code.

I agree with the downside that it hampers IDEs.

I think it's normal style, lots of projects have generated code.

**Comment by heinrich5991** [comment]
> The protocol generation has upsides, namely that one cannot accidentally write unsafe decoding code.

Also, it's useful to have a non-code description of the protocol for stuff like the Wireshark dissector.

**Comment by heinrich5991** [comment]
For the IDE use case, it might suffice to check in the generated code and let CI check that it's correct.

**Comment by Chairn** [comment]
BTW for IDE stuff, i added a symlink from build/src/game/generated into src/game/generated, that kinda helps for me.

**Comment by def-** [comment]
I like Python-generated code. But I also don‘t use an IDE, so don‘t notice the disadvantages. Doing this stuff straight in C++ at compile time can increase compile time and be more complex

**Comment by def-** [comment]
I always just used vim, sometimes with ctags and cscope for approximate jumps

**Comment by heinrich5991** [comment]
Why does the IDE not pick up the build directory's src directory btw? Is the IDE CMake-aware?

**Comment by C0D3D3V** [comment]
I'm also not against the python generator, it is a good way to build our unpacker/packer and datastructures. 

If we would build the structures based on inheritance at runtime, I guess it would not be as fast as it is now.  

It is just kind of strange in my eyes to generate sources, becuase you could do the same at runtime, at some costs but keep the same maintainability. 

If we would move the system in c++, it should be equally maintainable, if you would add a new netobj, it should be the same as if you would do it in phython, just add the definition, that should be it. If you add new sprites, or other assets, it should also only requiere one change as in phython.

**Comment by C0D3D3V** [comment]
Btw we could also think the other way around. For example for the variables definition for console commands and chat commands. If you add an command there, you have to recompile a lot of files. I guess if we would also generated them with python, it would maybe get faster. I'm not sure about it.

**Comment by heinrich5991** [comment]
> Btw we could also think the other way around. For example for the variables definition for console commands and chat commands. If you add an command there, you have to recompile a lot of files. I guess if we would also generated them with python, it would maybe get faster. I'm not sure about it.

I also thought about that. Thinking about it again, I don't quite see how we can tell CMake to collect all the config variables into one file, without rebuilding that file every build.

**Comment by Learath2** [comment]
The sixup part also relies on the ability to generate other protocols. I also don't mind the generation, but I don't use an IDE and my text editor seems to handle it just fine.

**Comment by Robyt3** [comment]
I tend to agree that generating the protocol code has more merits than downsides.

But generating `client_data.cpp/.h` and `server_data.cpp/.h` seems unnecessary. Those could be rewritten in a clean C++ way to build a collection of the client's resources. As explained in #9040, most of the server data is unused anyway. We should maybe also split the game properties that both client and server use from the resource handling that only the client needs.
